core/src/main/scala/kafka/network/SocketServer.scala

```diff
@@ -492,6 +455,64 @@ private[kafka] class Processor(val id: Int,
     }
   }
 
+  /* `protected` for test usage */
+  protected[network] def sendResponse(response: RequestChannel.Response) {
+    trace(s"Socket server received response to send, registering for write and sending data: $response")
+    selector.send(response.responseSend)
```

Comment:
Good catch @junrao. I think this can happen when connections are closed via `Selector.maybeCloseOldestConnection` for example. I am not sure it can happen for the case where a client closes a connection and we detect it via read/write (we mute the selector after processing a `completedReceive` and `poll` only returns one `completedReceive` per channel). Is my understanding correct?

I made the suggested change and added a test for it.